concrete-core/src/specification/engines/glwe_ciphertext_discarding_relinearization_computation.rs

```diff
@@ -0,0 +1,72 @@
+use super::engine_error;
+use crate::specification::engines::AbstractEngine;
+use crate::specification::entities::{GlweCiphertextEntity, GlweRelinearizationKeyEntity};
+
+engine_error! {
+    GlweCiphertextDiscardingRelinearizationError for GlweCiphertextDiscardingRelinearizationEngine@
+    PolynomialSizeMismatch => "The polynomial size of the input and output GLWE ciphertexts must \
+    be the same.",
+    InputGlweDimensionMismatch => "The GLWE dimension of the input ciphertext and key must be the \
+    same.",
+    OutputGlweDimensionMismatch => "The GLWE dimension of the output ciphertext is incorrect"
+}
+
+impl<EngineError: std::error::Error> GlweCiphertextDiscardingRelinearizationError<EngineError> {
+    pub fn perform_generic_checks<InputKey, InputCiphertext, OutputCiphertext>(
+        input_key: &InputKey,
+        input_ct: &InputCiphertext,
+        output: &OutputCiphertext,
+    ) -> Result<(), Self>
+    where
+        InputKey: GlweRelinearizationKeyEntity,
+        InputCiphertext: GlweCiphertextEntity,
+        OutputCiphertext: GlweCiphertextEntity<KeyDistribution = InputCiphertext::KeyDistribution>,
```

Comment:
Is there a way for us to use something like `<KeyDistribution = ---KeyEntity:Kind>`, then we can perform a check on both ciphertexts?

Reply:
Oh, in this situation we have to ensure that `InputCiphertext` and `OutputCiphertext` are both `GlweCiphertextEntity<KeyDistribution = InputKey::KeyDistribution>`

Reply:
Thanks, I'll make the change

Reply:
I think I've fixed this, I had to add a `KeyDistributionMarker` to `GlweRelinearizationKeyEntity`. I made a note that this should be considered as the secret distribution of the underlying `GlweSecretKeyEntity`s.